### PR TITLE
Implement Rust native article view as fallback on failed `article_parse_command`

### PR DESCRIPTION
**Pull request by 0x6b**
First of all, thank you for bringing this great TUI for us!

### Summary

Implement Rust native article view as fallback on failed `article_parse_command`.

### Context

Current implementation will display a link to the [`config.md`](https://github.com/aome510/hackernews-TUI/blob/main/docs/config.md#article-parse-command) file if the `article_parse_command` fails. Almost all of new users will see this error. Instead, we can use the `readable-readability` to provide similar capability as a fallback method. This might be more user friendly and nice default behavior. 

### Dependency Added

- [`readable-readability`](https://crates.io/crates/readable-readability)

### Screenshots

https://news.ycombinator.com/item?id=37214693

|`article_md`|New implementation|
|-|-|
|article_md 1|integrated 1|

Please note that it's not perfect for some pages as below, but I think it'd be acceptable.

https://news.ycombinator.com/item?id=25607320

|`article_md`|New implementation|
|-|-|
|aritcle_md 2|integrated 2|

### Final Words

Quick and not-so-scientific benchmarking shows that the native implementation is faster than the `article_md`, separate Node.js CLI, on my setup. This is just a proposal. I'm not sure if it's a good idea to add a new dependency to the project. Feel free to close this PR if you don't like it. If you like it, I'd add commits to update relevant documentation. Thank you!

### CODE REVIEW

File: hackernews_tui/src/client/mod.rs [hunk 1]
```diff
@@ -376,27 +376,65 @@ impl HNClient {
...
+                    .timeout(std::time::Duration::from_secs(config::get_config().client_timeout))
+                    .build()
+                    .get(url)
+                    .call()
+                    .map_err(|why| anyhow::anyhow!("failed to get url: {why}"))?
+                    .into_string()
+                    .map_err(|why| anyhow::anyhow!("failed to turn the response into string: {why}"))?;
+                let (nodes, metadata) = readable_readability::Readability::new()
+                    .base_url(url::Url::parse(url)
+                        .map_err(|why| anyhow::anyhow!("failed to parse url: {why}"))?)
```
Comment (aome510) [inline hackernews_tui/src/client/mod.rs:416]: can use `anyhow::Context`, e.g `.context("failed to parse url")` instead.

File: hackernews_tui/src/client/mod.rs [hunk 2]
```diff
@@ -376,27 +376,65 @@ impl HNClient {
...
             }
-        } else {
-            let stderr = std::str::from_utf8(&output.stderr)?.to_string();
-            Err(anyhow::anyhow!(stderr))
+            Err(_) => {
+                // fallback to the `readable-readability` crate if the command fails
+                let html = ureq::AgentBuilder::new()
+                    .timeout(std::time::Duration::from_secs(config::get_config().client_timeout))
+                    .build()
+                    .get(url)
```
Comment (aome510) [inline hackernews_tui/src/client/mod.rs:409]: This is a function inside `HNClient` which has a ureq client already. Can you update the function signature to take `&self` and use `self.client` instead?

File: hackernews_tui/src/client/mod.rs [hunk 3]
```diff
@@ -376,27 +376,65 @@ impl HNClient {
...
+                    .call()
+                    .map_err(|why| anyhow::anyhow!("failed to get url: {why}"))?
+                    .into_string()
+                    .map_err(|why| anyhow::anyhow!("failed to turn the response into string: {why}"))?;
```
Comment (aome510) [inline hackernews_tui/src/client/mod.rs:413]: can use `anyhow::Context` instead

File: hackernews_tui/src/client/mod.rs [hunk 4]
```diff
@@ -376,27 +376,65 @@ impl HNClient {
...
+                let (nodes, metadata) = readable_readability::Readability::new()
+                    .base_url(url::Url::parse(url)
+                        .map_err(|why| anyhow::anyhow!("failed to parse url: {why}"))?)
+                    .parse(&html);
+
+                let mut text = vec![];
+                nodes.serialize(&mut text)
+                    .map_err(|why| anyhow::anyhow!("failed to serialize nodes: {why}"))?;
```
Comment (aome510) [inline hackernews_tui/src/client/mod.rs:421]: can use `anyhow::Context` instead

File: hackernews_tui/src/client/mod.rs [hunk 5]
```diff
@@ -376,27 +376,65 @@ impl HNClient {
...
+                let title = metadata.page_title
+                    .or(metadata.article_title)
+                    .unwrap_or("(no title)".to_string());
+                let content = std::str::from_utf8(&text)
+                    .map_err(|why| anyhow::anyhow!("failed to turn the text into string: {why}"))?
```
Comment (aome510) [inline hackernews_tui/src/client/mod.rs:426]: can use `anyhow::Context` instead

**Review by aome510 — CHANGES_REQUESTED** [review]
> Quick and not-so-scientific benchmarking shows that the native implementation is faster than the article_md, separate Node.js CLI, on my setup. This is just a proposal. I'm not sure if it's a good idea to add a new dependency to the project. Feel free to close this PR if you don't like it. If you like it, I'd add commits to update relevant documentation. Thank you!

Looking good. Thank you for this PR. Can you update the documentation in https://github.com/aome510/hackernews-TUI/blob/main/docs/config.md#article-parse-command and also the error message in https://github.com/aome510/hackernews-TUI/blob/bfc9e7f01bfb9a268d9826c88e764c183fa5c699/hackernews_tui/src/view/async_view.rs#L55-L61?

File: hackernews_tui/src/client/mod.rs (hunk 1 above)
Comment (0x6b) [reply hackernews_tui/src/client/mod.rs]: Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

File: hackernews_tui/src/client/mod.rs (hunk 3 above)
Comment (0x6b) [reply hackernews_tui/src/client/mod.rs]: Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

File: hackernews_tui/src/client/mod.rs (hunk 4 above)
Comment (0x6b) [reply hackernews_tui/src/client/mod.rs]: Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

File: hackernews_tui/src/client/mod.rs (hunk 5 above)
Comment (0x6b) [reply hackernews_tui/src/client/mod.rs]: Addressed at [1d03f33](https://github.com/aome510/hackernews-TUI/pull/94/commits/1d03f33a403a36ded9f07de77d7e52186148f376).

File: hackernews_tui/src/client/mod.rs (hunk 2 above)
Comment (0x6b) [reply hackernews_tui/src/client/mod.rs]: Addressed at [28d89b5](https://github.com/aome510/hackernews-TUI/pull/94/commits/28d89b5a9f9394a8a5d0721f401363b04e8d6bfc)

**Comment by 0x6b** [comment]
@aome510 Thank you for the detailed review. I have addressed your comments. I think some updated wordings should be corrected per your liking, please treat as a start point. Welcome any further feedback. Thanks!

**Comment by 0x6b** [comment]
@aome510 Thanks! I should have applied `cargo fmt` before pushing commits, of course. Please review my recent commits and let me know if you need anything else.

**Comment by aome510** [comment]
Great! Thanks @0x6b

**Pull request merged**